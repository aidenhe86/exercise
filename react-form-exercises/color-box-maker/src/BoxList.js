import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
  const INITIAL_STATE = [
    { id: uuid(), backgroundColor: "red", width: 10, height: 10 },
    { id: uuid(), backgroundColor: "blue", width: 7, height: 8 },
    { id: uuid(), backgroundColor: "yellow", width: 10, height: 8 },
  ];
  const [boxes, setBoxes] = useState(INITIAL_STATE);
  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
  };
  const removeBox = (id) => {
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      {boxes.map((box) => (
        <Box
          key={box.id}
          id={box.id}
          backgroundColor={box.backgroundColor}
          width={box.width}
          height={box.height}
          handleRemove={removeBox}
        />
      ))}
    </div>
  );
};

export default BoxList;
