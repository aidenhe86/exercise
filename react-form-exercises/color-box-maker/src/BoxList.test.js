import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// function to act as input new box
function addBox(boxList, height = "5", width = "5", backgroundColor = "black") {
  const heightInput = boxList.getByLabelText("Height");
  const widthInput = boxList.getByLabelText("Width");
  const backgroundColorInput = boxList.getByLabelText("Background Color");

  fireEvent.change(heightInput, { target: { value: height } });
  fireEvent.change(widthInput, { target: { value: width } });
  fireEvent.change(backgroundColorInput, {
    target: { value: backgroundColor },
  });

  const btn = boxList.getByText("Add a new box!");
  fireEvent.click(btn);
}

// Smoke test
it("renders without crashing", function () {
  render(<BoxList />);
});

// Snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add new box", function () {
  const boxList = render(<BoxList />);

  //   have default 3 boxes
  expect(boxList.queryAllByText("X")).toHaveLength(3);

  //   add new box
  addBox(boxList);

  //   have 4 boxes
  expect(boxList.queryAllByText("X")).toHaveLength(4);
  expect(boxList.queryAllByText("X")[3]).toHaveStyle(`
    width: 5em,
    height: 5em,
    backgroundColor: black,
  `);

  //   form should be empty after submit
  expect(boxList.getAllByDisplayValue("")).toHaveLength(3);
});

it("can remove box", function () {
  const boxList = render(<BoxList />);

  //   have default 3 boxes
  expect(boxList.queryAllByText("X")).toHaveLength(3);
  expect(boxList.queryAllByText("X")[0]).toHaveStyle(`
    width: 10em,
    height: 10em,
    backgroundColor: red,
  `);

  fireEvent.click(boxList.getAllByText("X")[0]);

  //   should have 2 boxes
  expect(boxList.queryAllByText("X")).toHaveLength(2);
  expect(boxList.queryAllByText("X")[0]).toHaveStyle(`
    width: 7em,
    height: 8em,
    backgroundColor: blue,
  `);
});
