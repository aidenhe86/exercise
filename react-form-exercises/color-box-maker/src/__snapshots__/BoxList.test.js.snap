// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <button>
        Add a new box!
      </button>
    </form>
    <div
      class="Box"
      style="background-color: red; width: 10em; height: 10em;"
    >
      <button>
        X
      </button>
    </div>
    <div
      class="Box"
      style="background-color: blue; width: 7em; height: 8em;"
    >
      <button>
        X
      </button>
    </div>
    <div
      class="Box"
      style="background-color: yellow; width: 10em; height: 8em;"
    >
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
